src/lib/JIT.hs

```diff
@@ -405,52 +441,52 @@ compileInstr instr = case instr of
     let dt = scalarTy idt
     emitInstr dt $ L.BitCast x dt []
   ICall f args -> do
-    fImpName <- substM f
-    f' <- lookupImpFun fImpName
-    let ty@(IFunType cc argTys resultTys) = impFunType f'
-    fname <- case f' of
-      FFIFunction _ fname -> return fname
-      ImpFunction _ _ -> do
-        ~(Just (CFun cname _)) <- queryObjCache fImpName
-        -- TODO: track deps!
-        return cname
     args' <- mapM compileExpr args
-    let resultTys' = map scalarTy resultTys
-    case cc of
-      FFIFun -> do
-        ans <- emitExternCall (makeFunSpec fname ty) args'
-        return [ans]
-      FFIMultiResultFun -> do
-        resultPtr <- makeMultiResultAlloc resultTys'
-        emitVoidExternCall (makeFunSpec fname ty) (resultPtr : args')
-        loadMultiResultAlloc resultTys' resultPtr
-      CEntryFun -> do
-        exitCode <- emitInstr i64 (callInstr fTy fun args') >>= (`asIntWidth` i1)
-        compileIf exitCode throwRuntimeError (return ())
-        return []
-          where
-            fTy = funTy i64 $ map scalarTy argTys
-            fun = callableOperand (hostPtrTy fTy) $ topLevelFunName fname
-      CInternalFun -> do
-        exitCode <- emitExternCall (makeFunSpec fname ty) args' >>= (`asIntWidth` i1)
-        compileIf exitCode throwRuntimeError (return ())
-        return []
-      _ -> error $ "Unsupported calling convention: " ++ show cc
+    lookupSubstM f >>= \case
+      FunctionSubstVal f' lTy (IFunType cc _ _) -> do
+        case cc of
+          CEntryFun -> do
+            exitCode <- emitCallInstr lTy f' args' >>= (`asIntWidth` i1)
+            compileIf exitCode throwRuntimeError (return ())
+            return []
+          CInternalFun -> do
+            exitCode <- emitCallInstr lTy f' args' >>= (`asIntWidth` i1)
+            compileIf exitCode throwRuntimeError (return ())
+            return []
```

Comment:
Since those two are identical, perhaps you could do this:
```
case cc of 
  CEntryFun -> return ()
  CInternalFun -> return ()
  _ -> error "..."
exitCode <- emit ...
...
```

Reply:
Quite right. I didn't notice!